doc/pam_log.8
.\" This file is part of PAM-Modules -*- nroff -*-
.\" Copyright (C) 2001-2015 Sergey Poznyakoff
.\"
.\" PAM-Modules is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3, or (at your option)
.\" any later version.
.\"
.\" PAM-Modules is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with PAM-Modules.  If not, see <http://www.gnu.org/licenses/>.
.TH PAM_LOG 8 "March 28, 2014" "PAM-MODULES" "Pam-Modules User Reference"
.SH NAME
pam_log \- log messages to syslog
.SH SYNOPSIS
.nh
.na
\fBpam_log\fR [\fB\-audit\fR]\
 [\fB\-debug\fR[\fB=\fILEVEL\fR]]\
 [\fB\-noopen\fR]\
 [\fB\-waitdebug\fR[\fB=\fISECONDS\fR]]\
 [\fB\-pri=\fIFACILITY\fB.\fIPRIORITY\fR]\
 [\fB\-tag=\fILABEL\fR] \fITEXT\fR...
.ad
.hy
.SH DESCRIPTION
Sends diagnostic text to syslog.
.SH OPTIONS
.TP
\fB\-pri=\fIFACILITY\fB.\fIPRIORITY\fR
Send text to the given syslog facility and priority.  Valid values for
\fIFACILITY\fR are:
.BR user ,
.BR daemon ,
.BR auth ,
.BR authpriv ", and "
.BR local0 " through " local7 .
Valid values for \fIPRIORITY\fR are:
.BR emerg ,
.BR alert ,
.BR crit ,
.BR err ,
.BR warning ,
.BR notice ,
.BR info ", and "
.BR debug .

Either part can be omitted, in which case the following defaults are
used: \fIFACILITY=\fBauthpriv\fR, \fIPRIORITY=\fBinfo\fR.
.TP
\fB\-tag=\fILABEL\fR
Use \fILABEL\fR as the syslog tag, instead of the module name.
.TP
\fB\-debug\fR[\fB=\fINUMBER\fR]
Set debugging level (0 <= \fINUMBER\fR <= 100).
.TP
\fB\-audit\fR
Log full debugging information (equivalent to \fBdebug=100\fR).
.TP
\fB\-waitdebug=\fIN\fR
Wait for \fIN\fR seconds before starting up.  This option is intended
to facilitate attaching to the module with
.BR gdb (1).
It is available only if the package was configured with
the \fB\-\-enable\-debug\fR option.
.SH MODULE TYPES PROVIDED
All module types (\fBaccount\fR, \fBauth\fR, \fBpassword\fR and
\fBsession\fR) are provided.
.SH RETURN VALUES
.TP
.B PAM_IGNORE
Ignore this module.
.SH EXAMPLES
.EX
account requisite pam_log.so -tag FTP \\
      -pri=daemon.info User ${user:-unknown} is granted FTP access
.EE      
.SH NOTE
This manpage is a short description of \fBpam_log\fR.  For a detailed
discussion, including examples and usage recommendations, refer to the
\fBPAM-modules Manual\fR available in texinfo format.  If the \fBinfo\fR
reader and the tar documentation are properly installed on your
system, the command
.PP
.RS +4
.B info pam-modules
.RE
.PP
should give you access to the complete manual.
.PP
You can also view the manual using the info mode in
.BR emacs (1),
or find it in various formats online at
.PP
.RS +4
.B http://www.gnu.org.ua/software/pam-modules/manual
.RE
.PP
If any discrepancies occur between this manpage and the
\fBPAM-modules Manual\fR, the later shall be considered the authoritative
source.
.SH "SEE ALSO"
.BR pam.conf (5),
.BR pam.d (5),
.BR pam (8).
.SH AUTHORS
Sergey Poznyakoff <gray@gnu.org>
.SH "BUG REPORTS"
Report bugs to <bug\-pam\-modules@gnu.org.ua>.
.SH COPYRIGHT
Copyright \(co 2001-2014 Sergey Poznyakoff
.br
.na
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
.ad
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.\" Local variables:
.\" eval: (add-hook 'write-file-hooks 'time-stamp)
.\" time-stamp-start: ".TH [A-Z_][A-Z0-9_.\\-]* [0-9] \""
.\" time-stamp-format: "%:B %:d, %:y"
.\" time-stamp-end: "\""
.\" time-stamp-line-limit: 20
.\" end:



NEWS
pam-modules -- history of user-visible changes. 2015-01-30
Copyright (C) 2001, 2004-2005, 2007-2012, 2015 Sergey Poznyakoff
See the end of file for copying conditions.

Please send pam-modules bug reports to <bug-pam-modules@gnu.org.ua>


Version 1.9.90 (Git)

* pam_ldaphome reads LDAP configuration from /etc/ldap.conf

This is in addition to its regular configuration file.

* pam_ldaphome runs inirc-command with user privileges

To run the command with root privileges, the configuration
variable initrc-root must be set to true.

* New pam_ldaphome variable: user-keys-boundary

User key files can contain both keys managed by pam_ldaphome and
added by the user.  These two groups of keys must be separated by
a special comment line, which informs pam_ldaphome that all keys
below it must be retained.

This feature is enabled by the user-keys-boundary configuration
setting.  Its value defines a string which, when used  after a
'#' character, forms the delimiting comment.  E.g. if the
configuration file contains:

  user-keys-boundary :user

then the line '#:user' can be used to delimit ldap-synchronized
and user-specific keys.

Version 1.9, 2014-05-21

* New module pam_groupmember

Tests whether the user is a member of one or more groups.

* pam_ldaphome can invoke an external program

An external program defined with the inirc-command keyword is run
in the newly created user's home directory.  It can be used for
per-user customization of the files copied from the skeleton dir.

The examples directory contains a perl program "usergitconfig", which,
when used as inirc-command, initializes the user's .gitconfig file.

* New auxiliary utilities

** ldappubkey

The `ldappubkey' utility is a simple Perl program which takes user
login name as its argument and produces on the standard output public
ssh keys for that user, each on a separate line.  The program is
designed for use with `openssh' version 6.2p1 or higher.

** usergitconfig

Customizes user's `.gitconfig' file using attributes from his LDAP
entry.  This utility can be used with the initrc-command statement
in pam_ldaphome.conf file.

* Bugfixes


Version 1.8, 2013-07-29

* pam_ldaphome

This module creates the user home directory, if it does not
already exist, and updates his `.ssh/authorized_keys' file with the
keys from the LDAP database.

* pam_umotd

Pam_umotd displays a user-specific message of the day.  The text can
be taken either from a disk file, or read from the standard output of
a program launched for that purpose.  This module is Linux-specific.

* Bugfixes
** pam_fshadow made reentrant



Version 1.7, 2011-04-08

* Allow for use of `CALL proc' in MySQL queries.
* Minor bugfixes in pamck.


Version 1.6, 2009-02-25

* pamck

Pamck is a command line utility for checking PAM authentication and
other management groups.  E.g.:

   pamck -s login smith

attempts to authenticate user `smith' using PAM service name `login'.


Version 1.5, 2009-02-17

* Configure

New command line options:

  --disable-fshadow
  --disable-log
  --disable-regex

Improved autodetection of MySQL and PostgreSQL libraries.

Missing prerequisites for any module cause disabling of that module,
but the configuration process continues.


Version 1.4, 2008-03-20

* pam_mysql and pam_pgsql

** Session management

Session management is implemented for both modules.  Session
management queries are `session-start-query' and `session-stop-query'.

** Variable expansion in configuration file.

Old style of variable expansion has been dropped.  The `$name'
notation is used instead.  To convert your old configuration files,
replace %u with $user, and %p with $password.

** setenv-query

This new query allows to store arbitrary data in PAM environment.


Version 1.3, 2008-03-15

* pam_mysql and pam_pgsql

** Configuration file syntax

Long statements can be split over several lines by placing
'\' character at the end of each line.

** ldap passwords

Both modules understand passwords in LDAP form.  A special
configuration file statement `allow-ldap-pass' is provided to control
this feature.  By default, `allow-ldap-pass yes' is assumed.


Version 1.2, 2008-03-14

* Several fixes in debugging code and pam_mysql, pam_pgsql modules.

* pam_fshadow

By default extended regular expressions are used. 

* pam_regex transform=expr

New command line option `transform' allows to rewrite user names.


Version 1.1, 2007-08-11

* pam_fshadow allows to use virtual domains to specify alternate password
databases.  New options: regex, basic, extended, ignore-case, icase
and revert-index.

* pam_regex: ignore-case can be used as an alias to icase.

* New modules

pam_log                 Log arbitrary data
pam_mysql		Authenticate using a MySQL database
pam_pgsql		Authenticate using a PostgreSQL database


Version 1.0

	Added documentation, improved configuration suite.


Version 0.1

	Initial release. See README for short description.

^L
=========================================================================
Copyright information:

Copyright (C) 2001, 2004-2005, 2007-2015 Sergey Poznyakoff

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[  ]*$"
eval: (add-hook 'write-file-hooks 'time-stamp)
time-stamp-start: "changes. "
time-stamp-format: "%:y-%02m-%02d"
time-stamp-end: "\n"
end:




pam_log/pam_log.c
/* This file is part of pam-modules.
   Copyright (C) 2006-2008, 2010-2012, 2014-2015 Sergey Poznyakoff
 
   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 3 of the License, or (at your
   option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
 
   You should have received a copy of the GNU General Public License along
   with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include <graypam.h>


/* Command line parsing */
static int cntl_flags;
static long debug_level;

static int xargc;
static const char **xargv;
static int priority = LOG_INFO;
static int facility = LOG_AUTHPRIV;
static const char *syslog_tag = MODULE_NAME;
static int do_open = 1;

static struct keyword syslog_facility[] = {
        DCL("user",       LOG_USER),
	DCL("daemon",     LOG_DAEMON),
	DCL("auth",       LOG_AUTH),
	DCL("authpriv",   LOG_AUTHPRIV),
	DCL("local0",     LOG_LOCAL0),
	DCL("local1",     LOG_LOCAL1),
	DCL("local2",     LOG_LOCAL2),
	DCL("local3",     LOG_LOCAL3),
	DCL("local4",     LOG_LOCAL4),
	DCL("local5",     LOG_LOCAL5),
	DCL("local6",     LOG_LOCAL6),
	DCL("local7",     LOG_LOCAL7),
	{ NULL }
};

static struct keyword syslog_priority[] = {
        DCL("emerg",      LOG_EMERG ),
	DCL("alert",      LOG_ALERT ),
	DCL("crit",       LOG_CRIT ),
	DCL("err",        LOG_ERR ),
	DCL("warning",    LOG_WARNING ),
	DCL("notice",     LOG_NOTICE ),
	DCL("info",       LOG_INFO ),
	DCL("debug",      LOG_DEBUG ),
	{ NULL }
};

static int
parse_priority(struct pam_opt *opt, const char *str)
{
	int len;
	struct keyword *kw;

	for (len = 0; str[len]; len++)
		if (ispunct(str[len]))
			break;

	if (len) {
		kw = gray_find_keyword(syslog_facility, str, len);
		if (!kw) {
			_pam_log(LOG_ERR,
				 "unknown syslog facility: %*.*s",
				 len, len, str);
			return 1;
		}
		facility = kw->code;
	}
	
	if (str[len]) {
		str += len + 1;
		kw = gray_find_keyword(syslog_priority, str, strlen(str));
		if (!kw) {
			_pam_log(LOG_ERR,
				 "unknown syslog priority: %s", str);
			return 1;
		}
		priority = kw->code;
	}
	return 0;
}

struct pam_opt pam_opt[] = {
	{ PAM_OPTSTR(debug), pam_opt_long, &debug_level },
	{ PAM_OPTSTR(debug), pam_opt_const, &debug_level, { 1 } },
	{ PAM_OPTSTR(audit), pam_opt_const, &debug_level, { 100 } },
	{ PAM_OPTSTR(waitdebug), pam_opt_null, NULL, { 0 },
	  gray_wait_debug_fun },
	{ PAM_OPTSTR(tag), pam_opt_string, &syslog_tag },
	{ PAM_OPTSTR(pri), pam_opt_null, NULL, { 0 }, parse_priority },
	{ PAM_OPTSTR(open), pam_opt_bool, &do_open },
	{ NULL }
};


static void
_pam_parse(pam_handle_t *pamh, int argc, const char **argv)
{
	int i;
	const char **targv;

	gray_log_init(0, MODULE_NAME, LOG_AUTHPRIV);

	targv = gray_malloc(argc * sizeof (targv[0]));
	for (i = 0; i < argc; i++) {
		if (argv[i][0] == '-') {
			if (argv[i][1] == '-' && argv[i][2] == 0)
				break;
			targv[i] = argv[i] + 1;
		} else
			break;
	}

	gray_parseopt(pam_opt, i, targv);
	free(targv);
	
	xargc = argc - i;
	xargv = argv + i;

	closelog();
	gray_log_init(!do_open, syslog_tag, facility);
}

static int
echo(pam_handle_t *pamh, const char *prefix, int argc, const char **argv)
{
	char *str;
	gray_slist_t slist;

	_pam_parse(pamh, argc, argv);
	slist = gray_slist_create();
	if (prefix) {
		gray_slist_append(slist, prefix, strlen(prefix));
		gray_slist_append(slist, ": ", 2);
	}
	gray_expand_argv(pamh, xargc, xargv, slist);
	gray_slist_append_char(slist, 0);
	str = gray_slist_finish(slist);
	_pam_log(priority, "%s", str);
	gray_slist_free(&slist);
	return PAM_IGNORE;
}



PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

PAM_EXTERN int
pam_sm_chauthtok(pam_handle_t *pamh,int flags,int argc,const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

PAM_EXTERN int
pam_sm_acct_mgmt (pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

PAM_EXTERN int
pam_sm_open_session (pam_handle_t *pamh, int flags, int argc,
                     const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

PAM_EXTERN int
pam_sm_close_session (pam_handle_t *pamh, int flags, int argc,
		      const char **argv)
{
	gray_pam_init(PAM_IGNORE);
	return echo(pamh, __FUNCTION__, argc, argv);
}

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_log_modstruct = {
    "pam_log",
    pam_sm_authenticate,
    pam_sm_setcred,
    pam_sm_acct_mgmt,
    pam_sm_open_session,
    pam_sm_close_session,
    pam_sm_chauthtok,
};

#endif

/* end of module definition */
